game-app/game-core/src/main/java/games/strategy/triplea/ui/EditPanel.java

```diff
@@ -480,68 +483,95 @@ public void actionPerformed(final ActionEvent event) {
-                new PlayerChooser(getData().getPlayerList(), getMap().getUiContext(), false);
-            final JDialog dialog =
-                playerChooser.createDialog(getTopLevelAncestor(), "Select owner PUs to change");
-            dialog.setVisible(true);
-            final GamePlayer player = playerChooser.getSelected();
-            if (player == null) {
+
+            final Optional<GamePlayer> player = choosePlayer();
+            if (player.isEmpty()) {
               cancelEditAction.actionPerformed(null);
               return;
             }
-            getData().acquireReadLock();
-            final Resource pus;
-            try {
-              pus = getData().getResourceList().getResource(Constants.PUS);
-            } finally {
-              getData().releaseReadLock();
+
+            final Optional<Resource> resource = chooseResource();
+            if (resource.isEmpty()) {
+              cancelEditAction.actionPerformed(null);
+              return;
             }
-            if (pus == null) {
+
+            final Optional<Integer> newTotal = chooseResourceValue(player.get(), resource.get());
+            if (newTotal.isEmpty()) {
               cancelEditAction.actionPerformed(null);
               return;
             }
-            final int oldTotal = player.getResources().getQuantity(pus);
-            final JTextField pusField = new JTextField(String.valueOf(oldTotal), 4);
-            pusField.setMaximumSize(pusField.getPreferredSize());
+
+            final IEditDelegate delegate = EditPanel.this.frame.getEditDelegate();
+            final String result =
+                delegate.changeResource(player.get(), resource.get().getName(), newTotal.get());
+            if (result != null) {
+              JOptionPane.showMessageDialog(
+                  getTopLevelAncestor(),
+                  result,
+                  "Could not perform edit",
+                  JOptionPane.ERROR_MESSAGE);
+            }
+            cancelEditAction.actionPerformed(null);
+          }
+
+          private Optional<GamePlayer> choosePlayer() {
+            final PlayerChooser playerChooser =
+                new PlayerChooser(getData().getPlayerList(), getMap().getUiContext(), false);
+            final JDialog dialog =
+                playerChooser.createDialog(getTopLevelAncestor(), "Change Resources");
+            dialog.setVisible(true);
+            return Optional.ofNullable(playerChooser.getSelected());
+          }
+
+          private Optional<Resource> chooseResource() {
+            // Ignore VPS resources, since that's what the economy panel does.
+            final List<Resource> resources =
+                getData().getResourceList().getResources().stream()
+                    .filter(r -> !r.getName().equals(Constants.VPS))
+                    .collect(Collectors.toList());
+            if (resources.size() == 1) {
+              return Optional.of(resources.get(0));
```

Comment:
It's possible for there to be a null element in the resource list? This is surprising and seemingly painful.

Reply:
No. Optional.of() doesn't accept nullable, you're thinking .ofNullable().